elastic/integrations
# Fix: add missing fields for browser synthetics integration

## What does this PR do?

This PR fixes https://github.com/elastic/kibana/issues/123479.

It makes the waterfall chart page for a step to load properly because it adds missing fields. Those fields that were missing caused [this query](https://github.com/elastic/kibana/blob/f4046b7f56f78a8e547d2e80cdf09ecf7c6e455b/x-pack/plugins/uptime/public/components/monitor/synthetics/step_detail/use_step_waterfall_metrics.ts#L54) to return documents which had no `fields` property.

Therefore, we would end-up trying to read property from an undefined `fields` key [here](https://github.com/elastic/kibana/blob/f4046b7f56f78a8e547d2e80cdf09ecf7c6e455b/x-pack/plugins/uptime/public/components/monitor/synthetics/step_detail/use_step_waterfall_metrics.ts#L78).

## How to test this PR

1. Checkout this branch and make sure to build the package within `packages/synthetics` with `elastic-package build` (it's recommended that you also `elastic-package clean` before building).
2. Once built, you may have to copy this package to the directory in which your package registry finds packages (depending on how you're running it).
    In my case, I use the following command, for example: `cp -r ~/Repositories/integrations/build/integrations/synthetics ~/.elastic-package/profiles/default/stack/development/`.
    After copying the package, you may have to restart the package registry so that it picks up the new package.
3. Run `elastic-package install` within the `packages/synthetics` folder and that will install the new package. Make sure Kibana and the registry are running when you do that.
4. Create a new browser monitor using the Fleet UI (side panel > add integration). Add at least one step to that monitor, as in the example below.
    ```js
    step("example", async () => {
        await page.goto("https://www.elastic.co");
    });
    ```
5.  Wait for the monitor to run and access the monitors page, select your monitor, open its latest execution, and see the step's breakdown.

You should **not** see a crash as in the GIF below.

![broken_perf_breakdown](https://user-images.githubusercontent.com/6868147/150376334-71901e65-19af-4015-918c-6c9859e6b468.gif)

Please notice that this does _not_ happen for Heartbeat.

## Checklist



## Author's Checklist

As someone who hasn't as much experience with ES, I'd appreciate input on the following:

- [ ] Are my mappings correct? I've simply copied them from what we were using on Heartbeat
- [ ] Will the mappings apply just fine when we already have received data for indexes created using this index template? It didn't seem like they would. Is that a case we should worry about? What should we do with regards to this?
- [ ] Please also test E2E with an Elastic Agent.

## Related issues

This PR fixes https://github.com/elastic/kibana/issues/123479.


## Screenshots

Below, you'll see the chart working after the fix.

![waterfall_working](https://user-images.githubusercontent.com/6868147/150397970-ad7a0f37-8dac-499c-aa39-dc16324f5101.gif)


## Note

I'll open a further PR to make the hook for this chart in Kibana more resilient too, just so that it doesn't crash anyway if we don't find any relevant `fields`.
